design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,582 @@
+
+* **Status**: Draft
+* **Author**: [Alistair Doswald](https://github.com/AlistairDoswald)
+* **JIRAs**: [KEYCLOAK-9693](https://issues.jboss.org/browse/KEYCLOAK-9693), [KEYCLOAK-9694](https://issues.jboss.org/browse/KEYCLOAK-9694) and [KEYCLOAK-847](https://issues.jboss.org/browse/KEYCLOAK-847)
+
+## Motivation
+
+The [W3C Web Authentication - Two-Factor](https://github.com/Keycloak/Keycloak-community/blob/master/design/web-authn-two-factor.md) document specifies as sub tasks
+the administration of two authentication factors ([KEYCLOAK-9693](https://issues.jboss.org/browse/KEYCLOAK-9693)) and of multiple authenticators for users
+([KEYCLOAK-9694](https://issues.jboss.org/browse/KEYCLOAK-9694)).
+
+The purpose of this design document is to define how this is to be handled in the
+administration console, and therefore by extension to how Keycloak handles authentication
+on a logical level. The design may also specify functions, data structures and screen
+mockups as the design progresses.
+
+The design of the step up authentication ([KEYCLOAK-847](https://issues.jboss.org/browse/KEYCLOAK-847)) process is also defined in this proposal as it is linked to the multi-factor
+authentication (MFA) and the mechanisms that are defined in the one will have an impact
+on the other.
+
+The design also has the following constraints:
+* The design must be simple to use, but can be customised for more complex use cases. This is true for both the administrator when setting up the realm and the user when logging in with Keycloak.
+* In addition to two factor authentication, the design must take into account the use with passwordless authentication (i.e. a single factor authentication that is not a password).
+* The system must be easily extensible for groups that wish to add their own authenticators within Keycloak.
+
+For [KEYCLOAK-9693](https://issues.jboss.org/browse/KEYCLOAK-9693) this means changes
+to the authentication logic, changes to the authentication part of the admin console,
+changes to the authentication screens that the user sees during login, changes to the
+REST API, and changes to the database.
+
+For [KEYCLOAK-9694](https://issues.jboss.org/browse/KEYCLOAK-9694) this means
+changes to the *users > credentials* part of the administration console and to
+the REST API.
+
+For [KEYCLOAK-847](https://issues.jboss.org/browse/KEYCLOAK-847) this means changes
+to the authentication logic, changes to the authentication part of the admin console,
+changes to the client part of the admin console, extending Keycloak's comprehension
+of the OIDC and SAML protocols, and some small changes to the REST API.
+
+What this design proposal does _not_ cover is the specific handling of the WebAuthn
+and modifications to the account console, as other design proposals and JIRAs already
+specify those modifications.
+
+## Design proposal for [KEYCLOAK-9693](https://issues.jboss.org/browse/KEYCLOAK-9693)
```

Comment:
I'm not sure about the concepts you are proposing. Isn't the Authentication Flow enough to accommodate the necessary changes to enable MFA?

The Authentication Flow is what users configure to set up user authentication where these flows can be nested so that you can build more complex and dynamic flows. As such, I envision something like that when enabling MFA in Keycloak:

```
Auth type                         | Requirement                               | Actions
---------------------------------------------------------------------------------------
Cookie
MFA                                [X] ALTERNATIVE  [ ] REQUIRED  [ ] DISABLED  Actions >
  |- Username/Password             [X] REQUIRED
  |- OTP                           [ ] ALTERNATIVE  [X] REQUIRED  [ ] DISABLED
  |- Security Key                  [X] ALTERNATIVE  [ ] REQUIRED  [ ] DISABLED
  |- Passwordless                  [X] ALTERNATIVE  [ ] REQUIRED  [ ] DISABLED
```
The example above can be understood as:

* Users using a browser are automatically re-authenticated if a valid session is active on the server
* Users should use their username/password as a 1st factor and OTP as a 2nd factor, by default.
* Users are able to choose between two alternative authentication methods when at the OTP page: *Security Key* and *Passwordless* as alternative methods of authentication.

Another option would be to only ask the user for his username first, then show a page with the different authentication methods supported by the `MFA` authentication flow.

The `Actions` button on the `MFA` authentication flow redirects the admin to a configuration page where additional configuration could be set in regards to any custom setting we want to support for MFA.

I don't understand yet the need for the concepts you are proposing. I understand the motivation behind some of them such as the category, but I'm still not sure about it. For instance, in regards to users using multiple security keys:

* Isn't just a matter of allowing users to select the device they want to use ? Considering the user has properly registered the devices each one with a specific alias.

In case you are using "multiple-tokens":

* Considering that the admin has configured an authentication flow with both OTP and FIDO2 as 2nd factor, isn't just a matter of allowing the user to choose between these two options right after the first log in ?

Reply:
The problems I see that the current authentication flow is taken sequentially, and doesn't allow to switch between authentication categories. One of the reasons I suggest to make authentication factors explicit is to clarify the flow, and simplify the manner in which an administrator can define which type of credentials can be used at a specific part of the general authentication flow. Combined with authentication categories, it allows keycloak to know which credential options it can allow the user to present during the user's authentication at which step.

In the scheme you propose, if "passwordless" is for example a fido2 token with biometric identification, I don't know how an admin would decide that this is enough for two factors, or will decide only to set is as 2nd factor. Also, I may misunderstand what you're saying, but the way you describe your flow you can only choose passwordless after entering the username and password, which defeats the purpose. Also, I don't see in your flow any way to have an alternative to username password as a 1st factor. Currently it is required, but making explicit authentication factors would allow any alternative factor, say a biometric device as 1st factor, as long as it provides the necessary context for the following factors. Generally, by making authentication factors explicit, I believe that you can have this sort of flexibility.

The final advantage is that it can be easily tied in with the step-up, allowing to set an level on the authentication factor. It simplifies the matter for keycloak to know which set of authentication executions must be presented for the step up, since they are already bundled together.

You're right to suggest that authentication factors are more similar to a flow than to an executor however, and in the prototype I'm working on I'm currently extending authentication flow (I may refactor this later as I advance though, I've only just been able to start on prototyping this). The current default flow would anyway be needed to reworked to act as you suggest, both for allowing the user to switch between OTP and Fido2 (or any other 1st or 2nd factor), and for configuring the flow through the `Actions` button . In a way, what I suggest is exactly that: specialised flows that allow for alternative authentication executions to be set, with other benefits that are linked to their design.

Reply:
> The problems I see that the current authentication flow is taken sequentially, and doesn't allow to switch between authentication categories. One of the reasons I suggest to make authentication factors explicit is to clarify the flow, and simplify the manner in which an administrator can define which type of credentials can be used at a specific part of the general authentication flow. Combined with authentication categories, it allows keycloak to know which credential options it can allow the user to present during the user's authentication at which step.
> 
> In the scheme you propose, if "passwordless" is for example a fido2 token with biometric identification, I don't know how an admin would decide that this is enough for two factors, or will decide only to set is as 2nd factor. Also, I may misunderstand what you're saying, but the way you describe your flow you can only choose passwordless after entering the username and password, which defeats the purpose. Also, I don't see in your flow any way to have an alternative to username password as a 1st factor. Currently, it is required, but making explicit authentication factors would allow any alternative factor, say a biometric device as 1st factor, as long as it provides the necessary context for the following factors. Generally, by making authentication factors explicit, I believe that you can have this sort of flexibility.
>

Yeah, it is a sequential flow but this is something we can change in order to accommodate flows that rely on user preference as well as on the characteristics of the defined authentication methods.

My example is not clear about this but somehow we have an association between executions and their corresponding factor, this can be something that an admin can configure when at the configuration page of the flow (e.g.: after clicking Actions -> Config) or even by explicitly defining a factor to each execution definition in Keycloak.

Considering this, you could:

* Define which executions are required by default, sequentially, as a 1st and 2nd-factor authentication. In that case UsernamePassword + OTP
* Define which executions are alternatives as a 1st or 2nd factor. In that example, Passwordless (e.g.: FIDO+Bio) and Security Token are 1st and 2nd-factor alternatives, respectively. 

In an ideal flow, the login page would prompt the username (or show some avatar he can click). After that, the user is redirected to the 1st step which is set to prompt the user for his password but with options to select an alternative authentication method. In this case, the available options are "Passwordless".

After choosing one of the authentication methods in this 1st step the user is redirected to the 2nd step. In this step, the user is presented with OTP by default, but with options to select an alternative authentication method. In this case, Security Key.

However, there is a catch here that I'm not sure how you are addressing in your proposal. Considering the previous example where both UsernamePassword and Passwordless are defined as 1st-factor authenticators. How to configure that when users authenticate using Passwordless they don't need to inform a 2nd-factor ?

In a nutshell, I like your idea I'm just not sure if the best way to push the "Authenticator Factor" concept to much into the Flows but instead provide either more options in "Requirements" to configure the factors or, probably better, have a set of options in the Flow configuration that are specific for MFA (which I think gives much more flexibility in terms of configuration).
 
> The final advantage is that it can be easily tied in with the step-up, allowing to set an level on the authentication factor. It simplifies the matter for keycloak to know which set of authentication executions must be presented for the step up, since they are already bundled together.
> 

I agree with setting the level on the authentication factor/execution. But I think there some nuances that we need to discuss regarding that.

> You're right to suggest that authentication factors are more similar to a flow than to an executor however, and in the prototype I'm working on I'm currently extending authentication flow (I may refactor this later as I advance though, I've only just been able to start on prototyping this). The current default flow would anyway be needed to reworked to act as you suggest, both for allowing the user to switch between OTP and Fido2 (or any other 1st or 2nd factor), and for configuring the flow through the `Actions` button . In a way, what I suggest is exactly that: specialised flows that allow for alternative authentication executions to be set, with other benefits that are linked to their design.

Reply:
Another point I would like to make is regarding the association between the different authentication methods and whether or not they are considered a 1st, 2nd or Nth factor.

Taking as an example passwords and OTP, I think password is always a SFA while OTP will be always a 2FA. The same goes with a security key, which is always a 2nd factor.

Passwordless (e.g.: SMS, FIDO2+Bio, Notification) on the other hand can be considered a 1st factor as they don't require from users another factor.

Considering that, do we really need to distinguish in the UI the different mechanisms and whether or not they are a 1st or 2nd factor?

Reply:
What we have today is rather limiting, so I'm confident that we need to extend on that. I'm not yet convinced that authentication categories completely solves the problem though. Take WebAuthn for instance. That is a rather tricky one as it can act as a single factor, a 2nd factor or as multiple factors on its own, depending on the type of WebAuthn device used. Another example here is an external authenticator mechanism that can provide multiple factors, but is visible in Keycloak as a single authenticator.

Reply:
I am not sure if it is better to introduce Authentication factor or if we can just "enhance" Authentication execution and add the "LoA" to it (and maybe some other configuration options and tweaks).

Just few points:
- Migration: I think it is possible to change the default KC behaviour, however during migration from previous versions (EG. usecase when admin used KC 5 and now he is migrating to KC 7), the behaviour needs to be preserved.

- Some flows can be currently configured in a way that you can nest multiple alternative flows to more levels. For example default configuration of "First Broker Login" flow currently uses 3 levels. Not sure if this would be possible if you change ALTERNATIVE authentication executions to "authentication factors" as those are always 1st level if I understand correctly?

In other words, I think that support for ALTERNATIVE executions needs to be probably kept.

Reply:
> What we have today is rather limiting, so I'm confident that we need to extend on that. I'm not yet convinced that authentication categories completely solves the problem though. Take WebAuthn for instance. That is a rather tricky one as it can act as a single factor, a 2nd factor or as multiple factors on its own, depending on the type of WebAuthn device used. Another example here is an external authenticator mechanism that can provide multiple factors, but is visible in Keycloak as a single authenticator.

Yeah, it is. My take on this is that we need to rethink the authentication flow UIs in order to accommodate this proposal. Where MFA probably deserves its own set of UIs and (if we keep current authentication flow UIs) integration with the authentication flow UI to setup MFA to flows.

Yeah, WebAuthn API can be used to support different factors. But the factor also depends on the device you are using. I would say that some authenticators can be associated with more than one factor and admins should have the flexibility to use them in authentication flows accordingly with their requirements.

Reply:
> I am not sure if it is better to introduce Authentication factor or if we can just "enhance" Authentication execution and add the "LoA" to it (and maybe some other configuration options and tweaks).
> 
> Just few points:
> 
> * Migration: I think it is possible to change the default KC behaviour, however during migration from previous versions (EG. usecase when admin used KC 5 and now he is migrating to KC 7), the behaviour needs to be preserved.
> * Some flows can be currently configured in a way that you can nest multiple alternative flows to more levels. For example default configuration of "First Broker Login" flow currently uses 3 levels. Not sure if this would be possible if you change ALTERNATIVE authentication executions to "authentication factors" as those are always 1st level if I understand correctly?
> 
> In other words, I think that support for ALTERNATIVE executions needs to be probably kept.

IMO, and aligned with the points you made, we should keep existing behavior as much as possible without change it too much (even because I'm not sure if we need to in order to support MFA) but improve it to better support MFA.

Reply:
Yeah, something like that or even along the lines I proposed in the beginning. However, I'm almost confident that for the best is to provide a separated set of UIs for MFA configuration and then use the Authentication Flow UI to link a specific MFA config.

I'm also wondering if we could:

* Allow users to associate a LoA with each authentication execution (password, otp, security key, bio, etc)
* Allow users to configure the authentication executions that should kick in based on the LoA. For instance, Password is LoA `1` and OTP is defined to execute when User is authenticated and his security level is < `2`. Another example is Passwordless with LoA `3` which will not force the user to enter his OTP (> 2).
* Allow users to configure the LoA, thus the authenticators that should be executed based, based on different information about the context. For instance, if the user is authenticating from network XPTO he must follow the flow which provides LoA > 2.

I think the same idea could also be applied to setting LoA on the flows too. in a nutshell, the requested LoA will tell how the athentication will happen.

Reply:
Compared to @pedroigor's solution, I prefer the sub-flows idea, visually it imparts to the admin what is happening in what order. Before writing it in a 2nd draft though, I have some concerns about flow logic that I'd like to check.

(note that in the following text, when I say "actions" I mean either a flow or an execution)

Currently the flow logic is the following I believe:
* Process flow items sequentially
* Execute all required actions that are in the flow
* Execute all alternative actions that are in the flow, until the first one is valid, then skip all other alternative actions
* Skip all disabled actions
* Skip all optional flows, execute authentication executions if they are configured for the user

Sub-flows work the same way, but will result in either a success or a challenge for the main flow

With the proposed scheme, this would be modified in the following way:

* Process flow items sequentially. This is true most of the time, but not when you arrive in a situation where you have to choose between credentials, where keycloak must choose from the available executions in the subflow to 1) show credentials from all executions in the subflow and 2) show the page associated with the currently selected credential. This was solved in authentication factors, but for subflows I think that you must have a configurable option (in actions for example) that lets keycloak know that the logic is different?
* Execute all required actions that are in the flow
* Execute all alternative actions that are in the flow, until the first one is valid, then skip all other alternative actions. Again, except if the subflow is somehow marked to denote that the executions in the flow can be selected, in which case users can switch between all alternative actions
* Skip all disabled actions
* Process optional flows, IF the condition set is met, otherwise skip, execute authentication executions if they are configured for the user. I don't think an optional authentication execution makes any sense in flow in which one can choose between credentials though.

No changes are made to how subflows work.

To demonstrate with this logic with some examples, and to show the possible complexities of the solution, we would have:

```
Auth type                         | Options
---------------------------------------------------------------------------
Identity Provider Redirector       [x] Alternative  [ ] Required  [ ] Optional  
Auto                               [x] Alternative  [ ] Required  [ ] Optional  
  | - Cookie                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Kerberos                     [x] Alternative  [ ] Required  [ ] Optional  
Authenticate                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Username                     [ ] Alternative  [x] Required  [ ] Optional
  | - 1st                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - Password                [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
  | - 2nd                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - OTP                     [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
```
Note: The option "executions selectable" means that the execution logic presents the user with the choice of all his credentials that match the authentication categories of the authentication executions within the flow (basically, the user can choose which credential to use at that point).

Here we have a two factor authentication, and assuming we arrive to the authenticate subflow at 1st either password or webauthn. If password is selected we go to 2nd and can select between OTP and webauthn, but if we select webauthn, we are logged in as passwordless since webauthn is already successful from 1st.

```
Auth type                         | Options
---------------------------------------------------------------------------
Identity Provider Redirector       [x] Alternative  [ ] Required  [ ] Optional  
Auto                               [x] Alternative  [ ] Required  [ ] Optional  
  | - Cookie                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Kerberos                     [x] Alternative  [ ] Required  [ ] Optional  
Authenticate                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Username                     [ ] Alternative  [x] Required  [ ] Optional
  | - 1st                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - Password                [x] Alternative  [ ] Required  [ ] Optional
  | - 2nd                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - OTP                     [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
```
Same as before, except that we have password and 2nd factor

```
Auth type                         | Options
---------------------------------------------------------------------------
Identity Provider Redirector       [x] Alternative  [ ] Required  [ ] Optional  
Auto                               [x] Alternative  [ ] Required  [ ] Optional  
  | - Cookie                       [x] Alternative  [ ] Required  [ ] Optional  
Authenticate                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Username                     [ ] Alternative  [x] Required  [ ] Optional
  | - 1st                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - Kerberos                [x] Alternative  [ ] Required  [ ] Optional
       | - Password                [x] Alternative  [ ] Required  [ ] Optional
  | - 2nd                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - OTP                     [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
```
Similar to the first 2, except that kerberos is now an allowed 1st factor, but a 2nd factor is still required. However here, the user must explicitly select kerberos from the list.

```
Auth type                         | Options
---------------------------------------------------------------------------
Identity Provider Redirector       [x] Alternative  [ ] Required  [ ] Optional  
Auto                               [x] Alternative  [ ] Required  [ ] Optional  
  | - Cookie                       [x] Alternative  [ ] Required  [ ] Optional  
Authenticate                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Username                     [ ] Alternative  [x] Required  [ ] Optional
  | - 1st                          [ ] Alternative  [x] Required  [ ] Optional
       | - Kerberos                [x] Alternative  [ ] Required  [ ] Optional
       | - Password                [x] Alternative  [ ] Required  [ ] Optional
  | - 2nd                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - OTP                     [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
```
As before, except that now kerberos will automatically be tried as a 1st factor, then password will be presented if it's not present. I'm not sure how the case would be handled if we want kerberos, password or passwordless however with this logic. Probably by having the following:

```
Auth type                         | Options
---------------------------------------------------------------------------
Identity Provider Redirector       [x] Alternative  [ ] Required  [ ] Optional  
Auto                               [x] Alternative  [ ] Required  [ ] Optional  
  | - Cookie                       [x] Alternative  [ ] Required  [ ] Optional  
Authenticate                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Username                     [ ] Alternative  [x] Required  [ ] Optional
  | - 1st                          [ ] Alternative  [x] Required  [ ] Optional
       | - Kerberos                [x] Alternative  [ ] Required  [ ] Optional
       | - Subflow                 [x] Alternative  [ ] Required  [ ] Optional  -- Option set: executions selectable
             | - Password          [x] Alternative  [ ] Required  [ ] Optional
             | - WebAuthn          [x] Alternative  [ ] Required  [ ] Optional
  | - 2nd                          [ ] Alternative  [x] Required  [ ] Optional  -- Option set: executions selectable
       | - OTP                     [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
```

When conditional statements are introduced, this becomes a bit more complicated:

```
Auth type                         | Options
---------------------------------------------------------------------------
Identity Provider Redirector       [x] Alternative  [ ] Required  [ ] Optional  
Auto                               [x] Alternative  [ ] Required  [ ] Optional  
  | - Cookie                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Kerberos                     [x] Alternative  [ ] Required  [ ] Optional  
Authenticate                       [x] Alternative  [ ] Required  [ ] Optional  
  | - Username                     [ ] Alternative  [x] Required  [ ] Optional
  | - 1st                          [ ] Alternative  [ ] Required  [x] Optional  -- Option set: executions selectable
       | - Condition               [x] Required                                 -- Option set LoA < 1  
       | - Password                [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
       | - OnSuccess               [x] Required                                 -- Option set LoA = 1
  | - 2nd                          [ ] Alternative  [ ] Required  [x] Optional  -- Option set: executions selectable
       | - Condition               [x] Required                                 -- Option set LoA < 2 && RequestedLoA >= 2
       | - OTP                     [x] Alternative  [ ] Required  [ ] Optional
       | - WebAuthn                [x] Alternative  [ ] Required  [ ] Optional
       | - OnSuccess               [x] Required                                 -- Option set LoA = 2
```
I think this is what @stiant wants in terms of flows when LoA is injected. Here if the LoA of the user is less then 1 we enter the 1st flow (which will always be the case on 1st login), and the LoA is set to 1. The 2nd flow is skipped because `RequestedLoA` is not set.
When a client request a LoA = 2, the entire flow is re-executed:
* The cookie doesn't grant authentication because requested LoA is tested against the LoA of the user's session and is less
* 1st flow isn't executed because `LoA = 1`
* 2nd flow is executed because all conditions match. If the user had seleted passwordless with webauthn, it is automatically validated and the user goes to `LoA=2`, and otherwise the user can select between WebAuthn and OTP to do the step up.

I believe that this logic would match all the requirements for multifactor authentication. A criticism can be made that `Condition` and `On success` have to have some special construction as executors in order not to be considered as options by the `Option set: executions selectable`.

Also, even though this works, my team and I think that my initial proposal with authentication factors as specialised flows would be easier to configure and to use (and possibly to code), as they also allow all that is here. It would also be easier to allow backwards compatibility and avoid regression.

However conceptually it is the same, on one hand we create a specialised flow, and on the other we rewrite some of the behaviour of the existing default flow.

Reply:
Your authentication factor and sub flow are very similar concepts. I think now I understand better why you have this concept in your proposal. I can see now that your proposal introduces less changes or even no changes to the authentication flow as the "authentication factor" is basically a type of flow.

I'm still not sure about authentication category and if I understood the concept correctly, we could also solve the problem by creating a relationship between the executions and the credentials they support.

I'm almost convinced that we should consider creating this new type of flow  (authenticator factor) specifically for MFA configuration. I just don't like calling it 1st or N factor and explicitly push that to users. The flows are already in order of execution and it is up to the user to easily infer that by reading the UI. And as I mentioned before, executions are associated with a specific factor by nature (which is one of the principles in my proposal using only flows).

I agree that conceptually (maybe not in terms of terminology) we are all talking about the same thing though.

Reply:
> Here we have a two factor authentication, and assuming we arrive to the authenticate subflow at 1st either password or webauthn. If password is selected we go to 2nd and can select between OTP and webauthn, but if we select webauthn, we are logged in as passwordless since webauthn is already successful from 1st.

I think we should also consider the case when the user arrive to the authentication subflow (either 1st or 2nd) and one of the available authentication methods should be prompt, by default. This helps with UX if you know in advance that a specific authentication method is most commonly used.

Reply:
Yes, that's correct. The authentication factor is a specialised subflow which implicitly has the following characteristics:
* automatically executes sequentially all authentication executors that do not require user interaction
* then, if not validated, allows the user to choose between the remaining authentication executors
* Can have a LoA requirements set in its configuration page. 

The default action is actually already defined the the document (under **Authentication screens for the user**). The authentication method that is 1st prompted is the one of a credential that the user has selected as "preferred". If the user has none defined, it's the first that is in the list in the authentication factor.

And the main purpose of the authentication category _is_ to create the relationship between the the authentication executions and the credentials they support. The secondary purpose is to set other common elements between them, like the `amr` category. I've started on a 2nd draft of this proposal, and at @stianst's request I'll try to describe more explicitly what the authentication category is and how it's used and created.

Reply:
Having specialized authentication factor means hard-coding a lot of logic into the authentication processor. I can see it working for simpler use-cases we're thinking about now, but would not allow the needed flexbility to support other use-cases in the future. I'm not convinced about it at all and would still rather want to see the general concept of flow and sub-flows extended to accommodate new use-cases.

The flow of the logic for a sub-flow doesn't need to change much, I also don't think the auth category is needed.

Let's say an execution is extended to be able to provide the following meta-data:

* Is it the default selected by the user?
* Is it configured/enabled for a user?
* Some meta-data for human readable names etc.

Now the logic is when processor gets to a sub-flow with multiple execution alternatives in:

* If a specific execution is requested (through drop-down on login forms) render that one
* If a specific execution is configured as default for user render that one
* Otherwise render the first execution that is configured/enabled for the user

Reply:
I think we're aligned at this regard. The idea is to have flow/sub-flow with minimal changes to the authentication processor. 

There should be no specific authenticator factor flow, but extend the flow/executions to accommodate the new requirements.

This is pretty much what we start discussing in this review.

Reply:
OK, I think I see where we're going with this. Being able to select between executions simply replaces sequential execution when "selectable" authentication executions are available. If this is the case, then I think that the following logic for authentication executions makes sense:

* ALTERNATIVE for authentication execution denotes a selectable choice that can validate a subflow, and is the normal case.
* OPTIONAL for an authentication execution no longer makes sense in a multi-token, multi-factor setting, as it is superseded by the "select a credential" concept.
* REQUIRED is strange in a setting in which you can switch between authentication executions, as it implies that it must be done. Something like `Username` (selection of the user) has to be set to "required", but I'm not sure how you process the logic if other authentication executions can be set to "required".

I'm not sure what the logic should be if there's some non-user interactive authentication executions in the flow. Either they should always be executed first, or it's the responsibility of the administrator to make sure that the subflow makes sense, as otherwise, once a user-interactive authentication execution is encountered, they are no longer automatically executed.

The other case which is ambiguous is what happens if there there's a flow with subflows mixed with authentication executions, something like this:
```
Auth type
------------------------------------------------------------
Identity Provider Redirector
Auto                              [x] Alternative
  |- Cookie                       [x] Alternative
  |- Kerberos                     [x] Alternative
Authenticate                      [x] Alternative
  |- Username                     [x] Required
  |- 1st factor                   [x] Required
         |- Password              [x] Alternative
         |- Subflow 1
               |- ...
         |- Webauthn              [x] Alternative
         |- Subflow 2
               |- ...  
         |- OTP                   [x] Alternative
```
Then `Subflow 1` and `Subflow 2` will never be executed if they are set to **Alternative**, as they are not selectable. If they are set to **Required** (or **Optional** with conditions), then the other alternatives (`password`, `Webauthn` or `OTP`) are never executed, as `Subflow 1` and `Subflow 2` are a necessary condition to validate the 1st factor, no matter what they contain. So it's the responsibility of the administrator for making subflows that make sense. Does this seem correct?

Reply:
Yeah, I think so.

I see OPTIONAL as "if the user has the credential, then prompt him". So if the user does not have the credentials set, the corresponding choice won't' be shown. This is different than ALTERNATIVE, where you actually try to challenge the user regardless.

I see REQUIRED as "among these options, prompt for this credential first". So when the user lands at the login page, it will be shown the "required"/default challenge and the alternatives. Maybe REQUIRED could be changed to DEFAULT.

Reply:
I read the above example as 1st factor is required, with the following options password, subflow-1, webauthn, subflow-2 and otp.

Alternative as I interpret it should mean within a sub-flow there is a list of executions or sub-flows that are alternatives to each other where one has to be executed. They can be tried in order, or by user preference, or selected directly.

The most flexible would be if we allow executions and sub-flows to be alternatives to each other within a sub-flow. I can see cases where a step consists of multiple executions. For example Kerberos can collection username and authenticate, WebAuthn devices can contain username, but password requires username. Perhaps a bit fictional, but you could have something like:

[1st]
   [Kerberos]
   [WebAuthn]
   [Pass Flow]
      - [Username]
      - [Password]

Basically I want us to think about all options and crazy authentication schemes folks can come up with and make sure we design something that can be extended, rather than re-designed in the future.

Reply:
I find it slightly hard to interpret optional. I guess it means it is required if configured? I've never really liked it though. I would rather have a new requirement conditional, where the condition is clearly stated.

Reply:
Indeed, the OPTIONAL could be replaced by CONDITIONAL while still keeping the same behavior.

I don't quite understand your example but I think I got the idea. In a nutshell, administrators have the responsibility to use flows that make sense while we should do exactly what was defined (respecting order, requirements, etc). Like @AlistairDoswald suggested.

In his flow, we should still be able to provide selectable executions by looking at the executions within the subflows. Or even allow subflow selection by users, which I think would help in your case.